Unable to Build Docker 28.0.0 on x86 (32-bit)

As maintainer of the Alpine Linux docker packages, I am unable to compile from source on **x86** arch _(all other archs compile fine)_.

```
>>> docker: Unpacking /var/cache/distfiles/cli-28.0.0.tar.gz...
>>> docker: Unpacking /var/cache/distfiles/moby-28.0.0.tar.gz...
>>> docker: building engine
Removing bundles/
---> Making bundle: dynbinary (in bundles/dynbinary)
Building dynamic bundles/dynbinary-daemon/dockerd (linux/386)...
+ tee /builds/alpine/aports/community/docker/src/moby-28.0.0/go.mod
module github.com/docker/docker
go 1.22.0
+ trap 'rm -f "${ROOTDIR}/go.mod"' EXIT
+ GO111MODULE=on
+ GOTOOLCHAIN=local
+ go build -mod=vendor -modfile=vendor.mod -o bundles/dynbinary-daemon/dockerd -tags seccomp -buildmode=pie -ldflags '-w -X "github.com/docker/docker/dockerversion.Version=28.0.0" -X "github.com/docker/docker/dockerversion.GitCommit=af898abe44662d9554fb15ee4d4a7307f1b8e315" -X "github.com/docker/docker/dockerversion.BuildTime=2025-02-20T18:55:44.000000000+00:00" -X "github.com/docker/docker/dockerversion.PlatformName=" -X "github.com/docker/docker/dockerversion.ProductName=" -X "github.com/docker/docker/dockerversion.DefaultProductLicense="   ' -gcflags= github.com/docker/docker/cmd/dockerd
# github.com/docker/docker/libnetwork/drivers/bridge
libnetwork/drivers/bridge/port_mapping_linux.go:679:45: undefined: syscall.SYS_SETSOCKOPT
+ rm -f /builds/alpine/aports/community/docker/src/moby-28.0.0/go.mod
>>> ERROR: docker: build failed
```
https://github.com/moby/moby/blame/459686bbea46f0a50bc81850603dc399271c078b/libnetwork/drivers/bridge/port_mapping_linux.go#L679

It appears that Go does not support `syscall.SYS_SETSOCKOPT` on **linux/x86**.

https://github.com/golang/go/issues/19043#issuecomment-279195971 suggests...
> On 386 GNU/Linux you'll need to call something along the lines of `Syscall6(syscall.SYS_SOCKETCALL, 15, /* setsockopt arguments */)`
linux/386 multiplexes all socket syscalls into a one syscall (probably because there aren't enough registers to pass the arguments.)

@robmry ^^

--- a/libnetwork/drivers/bridge/port_mapping_linux.go
+++ b/libnetwork/drivers/bridge/port_mapping_linux.go
@@ -13,7 +13,6 @@ import (
 	"slices"
 	"strconv"
 	"syscall"
-	"unsafe"
 
 	"github.com/containerd/log"
 	"github.com/docker/docker/libnetwork/iptables"
@@ -675,14 +674,7 @@ func bindSCTP(cfg portBindingReq, port int) (_ portBinding, retErr error) {
 		syscall.SetsockoptInt(sd, syscall.IPPROTO_IPV6, syscall.IPV6_V6ONLY, 1)
 	}
 
-	options := sctp.InitMsg{NumOstreams: sctp.SCTP_MAX_STREAM}
-	if _, _, errno := syscall.Syscall6(syscall.SYS_SETSOCKOPT,
-		uintptr(sd),
-		sctp.SOL_SCTP,
-		sctp.SCTP_INITMSG,
-		uintptr(unsafe.Pointer(&options)), // #nosec G103 -- Ignore "G103: Use of unsafe calls should be audited"
-		unsafe.Sizeof(options),
-		0); errno != 0 {
+	if errno := setSCTPInitMsg(sd, sctp.InitMsg{NumOstreams: sctp.SCTP_MAX_STREAM}); errno != 0 {
 		return portBinding{}, errno
 	}
 

--- a/libnetwork/drivers/bridge/port_mapping_linux_386.go
+++ b/libnetwork/drivers/bridge/port_mapping_linux_386.go
@@ -1,0 +1,21 @@
+package bridge
+
+import (
+	"syscall"
+	"unsafe"
+
+	"github.com/ishidawataru/sctp"
+)
+
+const sysSetsockopt = 14 // See https://elixir.bootlin.com/linux/v6.13.3/source/include/uapi/linux/net.h#L40
+
+func setSCTPInitMsg(sd int, options sctp.InitMsg) syscall.Errno {
+	_, _, errno := syscall.Syscall6(syscall.SYS_SOCKETCALL, // See `man 2 socketcall`
+		sysSetsockopt,
+		uintptr(sd),
+		sctp.SOL_SCTP,
+		sctp.SCTP_INITMSG,
+		uintptr(unsafe.Pointer(&options)), // #nosec G103 -- Ignore "G103: Use of unsafe calls should be audited"
+		unsafe.Sizeof(options))
+	return errno
+}

--- a/libnetwork/drivers/bridge/port_mapping_linux_others.go
+++ b/libnetwork/drivers/bridge/port_mapping_linux_others.go
@@ -1,0 +1,21 @@
+//go:build linux && !386
+
+package bridge
+
+import (
+	"syscall"
+	"unsafe"
+
+	"github.com/ishidawataru/sctp"
+)
+
+func setSCTPInitMsg(sd int, options sctp.InitMsg) syscall.Errno {
+	_, _, errno := syscall.Syscall6(syscall.SYS_SETSOCKOPT,
+		uintptr(sd),
+		sctp.SOL_SCTP,
+		sctp.SCTP_INITMSG,
+		uintptr(unsafe.Pointer(&options)), // #nosec G103 -- Ignore "G103: Use of unsafe calls should be audited"
+		unsafe.Sizeof(options),
+		0)
+	return errno
+}

